### PR TITLE
Give hints about existential types appearing in error messages

### PR DESCRIPTION
**Pull request by lpw25**
This PR changes the names of existential types from things like `$Foo_'a` to just `$a` and instead adds an explicit hint describing where `$a` came from. For instance, this message from the testsuite:

```
 Line 2, characters 6-7:
 2 |   | A x
           ^
Error: This pattern matches values of type "$A_'a"
       The type constructor "$A_'a" would escape its scope
```
becomes:
```
 Line 2, characters 6-7:
 2 |   | A x
           ^
Error: This pattern matches values of type "$a"
       The type constructor "$a" would escape its scope
       Hint: "$a" is an existential type bound by the constructor "A".
```

It does this by broadening the `abstract_reason` type to a `type_origin` type that describes where a type has come from, and uses that to annotate existential types in the environment with the constructor they have come from.

The printing is done in the same way as the existing hints about conflicting names. Personally, I think these don't have the best indentation -- it would be better if the hint was unindented -- but I can't see an obvious way to achieve that and it's as much a problem with the conflict hints as with the new ones.

### CODE REVIEW

File: typing/printtyp.ml
Comment (gasche) [inline typing/printtyp.ml]: I would prefer if the matching was kept robust here.

File: typing/ctype.ml
Comment (gasche) [inline typing/ctype.ml]: I would prefer `Char.code 'a'` rather than `97`.

Maybe this could be reformulated as (untested):

```ocaml
let name =
  let letter = Char.chr (Char.code 'a' + !name_counter mod 26) in
  let num = !name_counter / 26 in
  if num = 0 then String.make 1 letter
  else Printf.sprintf "%c%d" letter num
```

File: typing/types.mli
Comment (gasche) [inline typing/types.mli]: I am already feeling tired of the `Origin_` prefix. Could we drop the disambiguation prefix?

```ocaml
and abstract_origin =
  | Definition  (* or Default? what did we mean exactly with 'def'? *)
  | Rec_check_regularity 
  | Existential of { cstr: string; }
```

**Comment by nojb** [comment]
Naïve question: do we need to introduce another kind of type variable, `$a`? Why not use `'a`?

**Review by gasche — COMMENTED** [review]
(Only very superficial review comments for now.)

**Comment by lpw25** [comment]
> Naïve question: do we need to introduce another kind of type variable, `$a`? Why not use `'a`?

The `$a` aren't a new kind of variable, they are just abstract types with generated names. Printing them as unification variables would be confusing because it would cause the compiler to say things like "`'a` cannot be unified with `int`" to the user.

Also worth pointing out that this PR isn't adding these generated abstract types with `$` in the name, it is just adjusting their names.

File: typing/ctype.ml
Comment (lpw25) [reply typing/ctype.ml]: I grabbed this logic from [printtyp.ml](https://github.com/ocaml/ocaml/blob/3afd70a70b9f8b305eabd2b06be76997718b0cd6/typing/printtyp.ml#L972). Would you prefer I update that code too, abstract the logic out into `misc.ml` and update it, or leave that part alone?

File: typing/printtyp.ml
Comment (lpw25) [reply typing/printtyp.ml]: In principle, I would prefer it not to be robust, because it's entire purpose is to pick exactly the cases marked with this constructor, and this constructor's entire purpose is to be matched by this case. There is no reasonable way to update the type with new constructors that would require adding more cases to this match.

I'm fine to update it though on the basis of some "always use robust matches" rule of thumb that is applied even when it is not actually beneficial -- I can easily be convinced it is better to have a simple rule and apply it universally.

File: typing/ctype.ml
Comment (gasche) [reply typing/ctype.ml]: I would certainly prefer if we shared the code somewhere instead of duplicating it, and if the code was as nice as reasonably possible, so I vote Misc.

Note: Printtyp is careful to avoid conflicts in the environment -- there is a counter but also a separate freshness check. Do I correctly understand that we don't need to avoid conflicts with user names because existential names are a separate namespace in which users cannot create names? (This might be worth a small comment somewhere.)

File: typing/printtyp.ml
Comment (gasche) [reply typing/printtyp.ml]: I disagree with your analysis in the first paragraph. Currently `rec_check_regularity` is the only case of dummy names that we handle in this way, but I could very well see other type-checking processes injecting dummy abstract types for different reasons, and expecting to be asked to update the current function with a specialized error message.
(But: it may be that I don't understand the usage context for the somewhat arcane `warn_on_missing_def`, and that the above is unlikely due to a very restricted usage scenario.)

File: typing/ctype.ml
Comment (gasche) [inline typing/ctype.ml:2540]: I don't understand why we need to reuse `type_origin` here. This comes in play when I am turning an abstract type `type a` into an abbreviation `type a = b` where `b` is itself abstract. Why is it necessary to give `a` the same origin as `b` in this case? I would assume that the abbreviation will be expanded in the contexts than need it, and then the type-checker will work with `b` and access its origin directly.

**Comment by gasche** [comment]
I like the feature, and the code is pleasantly not-too-invasive. If we had a type-checker expert at hand they should review the design, but if they are busy I would feel confident approving after a review for code quality -- I have done a pass already and don't expect that there is much more for me to discuss.

File: typing/ctype.ml
Comment (lpw25) [reply typing/ctype.ml:2540]: We are giving `type a = b` the same origin as the original `type a` definition, not the origin of `b`.

Comment (gasche) [reply typing/ctype.ml:2540]: Ah, but in this context we know already that `a` is abstract, so we don't really need the funny `type_origin : declaration -> type_origin` function.

Looking at the code I understand that it would be invasive to collect the origin at the point where we check that `a` is abstract (I think this is in `is_instantiable env path` called by the callers of `add_gadt_equation`, let's wait for view patterns before we turn it into a data-returning function), but maybe a local match that asserts that the type is abstract would be simpler and clearer?

Comment (lpw25) [reply typing/ctype.ml:2540]: That function isn't "funny": all type declarations have an origin, but only abstract declarations have an interesting one. If we did try to add equations to record or variant types in the future, then we would still want to propagate their (boring) origins.

Comment (gasche) [reply typing/ctype.ml:2540]: I still find it funny, but you know the code better than I do so I will bet that you are more in the right. Resolved.

File: typing/ctype.ml
Comment (Octachron) [reply typing/ctype.ml]: Name conflicts will be resolved later in `Printtyp` between variable names, this part of the code is choosing alphabetically increasing variable names for anonymous existential types in constructor definitions:
```
type t = X: _ * _ * _ -> t
let error (X(a,_,c)) (X(_,_,c1)) = X([c;c1],a,a);;
```
>Error: This expression has type $c1 but an expression was expected of type $c
       Hint: $c and $c1 are existential types bound by the constructor X.

File: typing/ctype.ml
Comment (Octachron) [inline typing/ctype.ml:2534]: Do we want to document that `add_gadt_equation` requires the precondition  "`source` is available in `uenv`" (which is satisfied in all three call sites)?

Comment (lpw25) [reply typing/ctype.ml:2534]: I kind of think that the default assumption is that a function which takes an env and a type should provide an env that contains the type. If you prefer I could default to `Origin_definition` or even the yet-to-exist `Origin_other` which I was considering adding for places like `reify` that don't yet have their own origin but aren't definitions either.

File: testsuite/tests/typing-gadts/unexpected_existentials.ml
Comment (Octachron) [inline testsuite/tests/typing-gadts/unexpected_existentials.ml]: This changes is due to the code handling "unexpected" existentials at `typecore.ml` is looking for a existential which is not named "$Constr" before switching between
```ocaml
but the constructor "..." introduces existential types.
```
or 
```ocaml
but this pattern introduces the existential type "...".
```
This part of the error message needs to be simplified along this PR. I propose to keep only the constructor form.

File: typing/ctype.ml
Comment (Octachron) [reply typing/ctype.ml:2534]: I was more thinking about the `missing cmi` case: most of the other calls to `find_type` in `ctype.ml` handle this case directly, but `add_gadt_equation` can elide it since it has already been filtered by `is_intantiable`. 

I wouldn't mind an `Origin_other` for `reify`, but the `assert false` seems more self-documenting here.
(I am fine without adding a comment, this was mostly a review note for myself).

**Comment by Octachron** [comment]
I tend to agree that this change overall reduces noise:

On the one hand, we lose some information locality because the name `$a` no longer carries the constructor name.
On the other hand, with meaningful constructor argument names or locally named existentials, the constructor names don't add much more information , but they add a lot of redundancy.
Keeping the origin of the existentials in a shared hint sounds like a good compromise.
The fact that the hint explains what the strange `$a` means is also an improvement.

I am wondering if later we could show the constructor definition in the hint to gives an even more useful context.
But that would require hints to be a distinct part of the error report rather than ad-hoc part of the error message, which could be a way to solve the indentation issue (but that should be a matter for another PR).
 
A small remark, the manual description of the existential nomenclature should be updated. In particular, we are left with only three categories now that we always name anonymous constructor argument types.

**Comment by lpw25** [comment]
Comments addressed and Changes entry added

**Review by gasche — APPROVED** [review]
I find the revised implementation (and documentation) satisfactory. Approved.

I would think of "squashing" the commits together as they are not split in independent changes currently.

**Comment by garrigue** [comment]
A tiny amount of information is lost here: originally the names combined the constructor name, and the name of the existential variable inside the definition. Here the hint only gives the constructor name.
Of course this only matters if the same constructor introduces several existentials.
It should be possible to add the name of the variable to `type_origin`, and include it in the hint. Or would it be confusing ?

**Comment by lpw25** [comment]
We could add it to give an error like:
```
Hint: "$a" is an existential type "'a" bound by the constructor "A".
```
normally it would be pretty redundant, but I guess in some cases it would be more like:
```
Hint: "$a3" is an existential type "'a" bound by the constructor "A".
```
where the two names aren't quite identical, and so perhaps useful.

**Comment by gasche** [comment]
How easy would it be to always reuse the variable name as prefix, with the counter as suffix?

**Comment by Octachron** [comment]
That's already what happens: we use the variable name in the constructor definition if it exists when we choose the existential variable name at instantiation, and the type printer only adds a numerical suffix in case of collisions:
```ocaml
type any = Any: 'any * 'more -> any
type other = Other: 'any * 'more -> other
type variant = Variant: 'any * 'more -> variant
let f (Any(x,y)) (Other(a,_)) (Variant(r,_)) = (x,y) = (r,a)
``` 
> Error: This expression has type $any2 but an expression was expected of type
         $any
       Hint: $any is an existential type bound by the constructor Any.
       Hint: $any2 is an existential type bound by the constructor Variant.

**Comment by gasche** [comment]
Ah, so I don't think that any change is necessary?

**Comment by garrigue** [comment]
I see, since all the variables were named `$a` I thought that the name was forgotten.
I still think that the connection would be clearer if the existential was named `$'a` rather than `$a`.

**Pull request merged**
